File: market-application/src/main/java/com/example/market/application/service/RegisterProductService.java
package com.example.market.application.service;

import com.example.market.application.command.RegisterProductCommand;
import com.example.market.application.port.in.RegisterProductUseCase;
import com.example.market.application.port.out.ProductRepository;
import com.example.market.domain.catalog.Product;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;

@Service
@RequiredArgsConstructor
@Slf4j
public class RegisterProductService implements RegisterProductUseCase {

    private final ProductRepository products;
    private final Clock clock;

    @Override
    @Transactional
    public Product register(RegisterProductCommand cmd) {
        Product product = Product.create(
                cmd.brand(), cmd.modelName(), cmd.styleCode(),
                cmd.category(), cmd.releaseDate(), cmd.imageUrl(), clock.instant());
        products.save(product);
        log.info("product registered id={} brand={} model={}",
                product.id(), product.brand(), product.modelName());
        return product;
    }
}

File: market-application/src/main/kotlin/com/example/market/application/service/RegisterProductService.kt
package com.example.market.application.service

import com.example.market.application.command.RegisterProductCommand
import com.example.market.application.port.`in`.RegisterProductUseCase
import com.example.market.application.port.out.ProductRepository
import com.example.market.domain.catalog.Product
import java.time.Clock
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
open class RegisterProductService(
    private val products: ProductRepository,
    private val clock: Clock,
) : RegisterProductUseCase {

    @Transactional
    override fun register(command: RegisterProductCommand): Product {
        val product = Product.create(
            command.brand, command.modelName, command.styleCode,
            command.category, command.releaseDate, command.imageUrl, clock.instant(),
        )
        products.save(product)
        log.info(
            "product registered id={} brand={} model={}",
            product.id, product.brand, product.modelName,
        )
        return product
    }

    companion object {
        private val log = LoggerFactory.getLogger(RegisterProductService::class.java)
    }
}
